File: FinLib.NET/FinLib/Portfolio/Strategy.cs
using System;
using FinLib.Options;

namespace FinLib.Portfolio;

public class Strategy: IDisposable, IProfit<double>
{
    private readonly unsafe Strategy_native* _handle;
    internal unsafe Strategy_native* GetPtr() => _handle;

    public Strategy()
    {
        unsafe
        {
            _handle = NativeMethods.strategy_new();
        }
    }

    public nuint Size
    {
        get
        {
            unsafe
            {
                return NativeMethods.strategy_size(_handle);
            }
        }
    }

    public void Add(IProfit<double> component)
    {
        unsafe {
            if (component is OptionContract oc){
                NativeMethods.strategy_add_option_component(_handle, oc.GetPtr());
            }
            else if (component is Swap.Swap s)
            {
                NativeMethods.strategy_add_swap_component(_handle, s.GetPtr());
            }
            else
            {
                throw new NotImplementedException();
            }
        }
    }

    public double Payoff(double underlying)
    {
        unsafe
        {
            return NativeMethods.strategy_payoff(_handle, underlying);
        }
    }

    public double Profit(double underlying)
    {
        unsafe
        {
            return NativeMethods.strategy_profit(_handle, underlying);
        }
    }

    private void ReleaseUnmanagedResources()
    {
        unsafe
        {
            NativeMethods.strategy_destroy(_handle);
        }
    }

    public void Dispose()
    {
        ReleaseUnmanagedResources();
        GC.SuppressFinalize(this);
    }

    ~Strategy()
    {
        ReleaseUnmanagedResources();
    }
}

File: finlib-cpp/src/finlib/portfolio/strategy.cpp
//
// Created by Andy Pack on 19/06/2025.
//

#include "strategy.hpp"

namespace finlib {
    Strategy::Strategy() {
        handle = finlibrs::strategy_new();
    }

    void Strategy::add_component(std::unique_ptr<OptionContract> component) {
        finlibrs::strategy_add_option_component(handle, component->handle);
    }

    void Strategy::add_component(std::unique_ptr<Swap> component) {
        finlibrs::strategy_add_swap_component(handle, component->handle);
    }

    double Strategy::payoff(double underlying) {
        return finlibrs::strategy_payoff(handle, underlying);
    }

    double Strategy::profit(double underlying) {
        return finlibrs::strategy_profit(handle, underlying);
    }

    size_t Strategy::size() {
        return finlibrs::strategy_size(handle);
    }

    Strategy::~Strategy() {
        finlibrs::strategy_destroy(handle);
    }
}
